design/aws-ecs/aws-ecs-service-addons.md

```diff
@@ -0,0 +1,108 @@
+- __Extensions are aware of each other__: There are no hidden "gotchas" when enabling a service extension. Extensions know of each other's existence. For example the application's container is aware that there is an Envoy proxy and that it must wait until the Envoy proxy has started before the app starts.
+- __Service class is agnostic to EC2 vs Fargate__: Rather than choosing up the capacity strategy in the class name customers just create a service, and add it to an environment. The service automatically modifies its settings as needed to adjust to the capacity available in the environment.
+
+
+## Implementation
+
+The new `Service` class acheives its goals via a method `Service.add()` This method is used to add extensions to the service. Each extension is responsible for customizing the service to support a specific feature. The following example shows how the end user instantiates a service with various feature's enabled:
+
+```js
+const myService = new Service(stack, 'my-service', {});
+myService.add(new Container({
+  cpu: 1024,
+  memoryMiB: 2048,
+  trafficPort: 80,
+  image: ContainerImage.fromRegistry('nathanpeck/greeter')
+}));
+myService.add(new AppMeshAddon({ mesh }));
+myService.add(new FireLensAddon());
+myService.add(new XRayAddon());
+myService.add(new CloudwatchAgentAddon());
+myService.add(new HttpLoadBalancerAddon());
+```
+
+Each addon is responsible for one aspect of the resulting service. The final generated service is the sum of all the features that were enabled. There are a few benefits to this approach:
+
+- Users no longer have to select their full architecture up front as with the current ECS patterns. Instead of chosing between `ApplicationLoadBalancedEc2Service` and `NetworkLoadBalancedEc2Service` up front, they can actually deploy a service, and then later add or remove an `HttpLoadBalancerAddon` from the service without needing to change the base class in their code.
+- Users can enable any number of different combinations of features which were not accessible before. They can build the same load balanced services they could before, but now they can also enable App Mesh, X-Ray, Cloudwatch Agent, and FireLens on the service at the same time.
+- The code for each extension is its own self contained class. Rather than bloating the ECS constructs with logic that is specific to another service all of that logic can live in the extension itself.
+
+## Service Lifecycle
+
+The new `Service` construct differs a bit from other CDK L2+ constructs. Most CDK constructs are fully assembled in the constructor and are then immutable. The `Service` class is not like this. The constructor does nothing but initialize an empty list of extensions. Users add extensions to this list, to define their intention about what should be built later on. The service constructs do not actually exist at this point. The only thing that exists is a list of user specified intentions about what constructs they do want to exist.
```

Comment:
I'd love to hear from @eladb, @RomainMuller  or @rix0rrr or anyone from the core team about using this pattern for constructing a resource and what implications it might have in the JSII.